### PR TITLE
[Refactor] Add ColumnIteratorDecorator to reduce code duplication

### PR DESCRIPTION
**Pull request by sduzh**
This patch was splitted from another bigger PR #43545

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by sduzh** [comment]
@Mergifyio backport branch-3.3

File: be/src/storage/rowset/column_iterator.cpp [hunk 1]
```diff
@@ -64,4 +64,28 @@ Status ColumnIterator::fetch_dict_codes_by_rowid(const Column& rowids, Column* v
     return fetch_dict_codes_by_rowid(p, rowids.size(), values);
 }
 
+Status ColumnIterator::next_batch(const SparseRange<>& range, Column* dst) {
```
Comment (sduzh) [inline be/src/storage/rowset/column_iterator.cpp:67]: Add default implementation of `next_batch`, the original implementation was return `NotSupported` error.

File: be/src/storage/rowset/column_iterator.cpp [hunk 2]
```diff
@@ -64,4 +64,28 @@ Status ColumnIterator::fetch_dict_codes_by_rowid(const Column& rowids, Column* v
...
+        auto r = Range<>{iter.next(to_read)};
+        auto n = size_t{r.span_size()};
+        RETURN_IF_ERROR(next_batch(&n, dst));
+        CHECK_EQ(r.span_size(), n);
+        to_read -= n;
+    }
+    return Status::OK();
+}
+
+Status ColumnIterator::fetch_values_by_rowid(const rowid_t* rowids, size_t size, Column* values) {
```
Comment (sduzh) [inline be/src/storage/rowset/column_iterator.cpp:81]: ditto

File: be/src/storage/rowset/column_iterator_wrapper.h [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+        return _iter->next_batch(range, dst, path);
+    }
+
+    Status fetch_subfield_by_rowid(const rowid_t* rowids, size_t size, Column* values) override {
+        return _iter->fetch_subfield_by_rowid(rowids, size, values);
+    }
+
+protected:
+    ColumnIterator* _iter;
+    Ownership _ownership;
```
Comment (kevincai) [inline be/src/storage/rowset/column_iterator_wrapper.h:107]: `_ownership` can be eliminated, use `unique_ptr` to explicitly manages the lifecycle and ownership.

File: be/src/storage/rowset/column_iterator_wrapper.h [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+
+#pragma once
+
+#include "common/macros.h"
+#include "common/ownership.h"
+#include "storage/rowset/column_iterator.h"
+
+namespace starrocks {
+
+class ColumnIteratorWrapper : public ColumnIterator {
```
Comment (kevincai) [inline be/src/storage/rowset/column_iterator_wrapper.h:23]: this is the Decorator design pattern, name it as FilterColumnIterator or something might be explicit than `XXXWrapper`, as the FilterInputStream/FilterOutputStream has been well-known and easy to understand.

File: be/src/storage/rowset/column_iterator_wrapper.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+public:
+    explicit ColumnIteratorWrapper(ColumnIterator* iter, Ownership ownership) : _iter(iter), _ownership(ownership) {}
```
Comment (kevincai) [inline be/src/storage/rowset/column_iterator_wrapper.h:25]: provide a `parent()` interface for accessing its original ColumnIterator directly.

File: be/src/storage/rowset/column_iterator_wrapper.h (hunk 3 above)
Comment (sduzh) [reply be/src/storage/rowset/column_iterator_wrapper.h:107]: The ownership is non-deterministic.

Comment (kevincai) [reply be/src/storage/rowset/column_iterator_wrapper.h:107]: I mean the unique_ptr can be initialized based on the ownship passed in constructor, so in destructor, you don't have to do anything, everything is settled when the contruction done.

File: be/test/storage/rowset/series_column_iterator_test.cpp [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+#include "storage/rowset/series_column_iterator.h"
+
+#include <gtest/gtest.h>
+
+#include "column/fixed_length_column.h"
+#include "testutil/assert.h"
+
+namespace starrocks {
+
+TEST(SeriesColumnIteratorTest, test01) {
```
[the remark this answers is not in the document]
Comment (sduzh) [reply be/test/storage/rowset/series_column_iterator_test.cpp]: Updated

File: be/src/storage/rowset/column_iterator_wrapper.h (hunk 3 above)
Comment (sduzh) [reply be/src/storage/rowset/column_iterator_wrapper.h]: Updated

File: be/src/storage/rowset/column_iterator_wrapper.h (hunk 4 above)
Comment (sduzh) [reply be/src/storage/rowset/column_iterator_wrapper.h]: Updated

File: be/src/storage/rowset/column_iterator_wrapper.h (hunk 5 above)
Comment (sduzh) [reply be/src/storage/rowset/column_iterator_wrapper.h]: Updated

**Review by kevincai — DISMISSED** [review]

**Comment by sduzh** [comment]
@kevincai PTAL again, force pushed to resolve clang-format error.

**Review by kevincai — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Pull request merged**